### PR TITLE
APIv4 - CiviCase API: Fix openening a case with current user as creator

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Fixes APIv4 to be able to open a Case with the current user as creator.

Before
----------------------------------------
Passing `user_contact_id` as the case creator or case contact causes an error.

After
----------------------------------------
Works.

Technical Details
----------------------------------------
The `DAOActionTrait::writeObjects` function was formatting values but not by reference, so the CiviCase `writeObjects` function was using unformatted values to open the case, which would contain the raw string `user_contact_id` instead of the processed value.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@colemanw 

Stacktrace
api\v4\Action\BasicCustomFieldTest::testWithSingleField
Error: Cannot pass parameter 1 by reference

/home/jenkins/bknix-dfl/build/core-20238-1pay5/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOUpdateAction.php:63
/home/jenkins/bknix-dfl/build/core-20238-1pay5/web/sites/all/modules/civicrm/Civi/Api4/Provider/ActionObjectProvider.php:68
/home/jenkins/bknix-dfl/build/core-20238-1pay5/web/sites/all/modules/civicrm/Civi/API/Kernel.php:149
/home/jenkins/bknix-dfl/build/core-20238-1pay5/web/sites/all/modules/civicrm/Civi/Api4/Generic/AbstractAction.php:239
/home/jenkins/bknix-dfl/build/core-20238-1pay5/web/sites/all/modules/civicrm/tests/phpunit/api/v4/Action/BasicCustomFieldTest.php:75
/home/jenkins/bknix-dfl/extern/phpunit7/phpunit7.phar:615

File: Civi/Api4/Generic/Traits/CustomValueActionTrait.php [hunk 1]
```diff
@@ -51,7 +51,7 @@ public function getEntityName() {
   /**
    * @inheritDoc
    */
-  protected function writeObjects($items) {
+  protected function writeObjects(&$items) {
     $fields = $this->entityFields();
     foreach ($items as $idx => $item) {
```
Comment (aydun) [inline Civi/Api4/Generic/Traits/CustomValueActionTrait.php:56]: Should `$items as $idx => $item` be `$items as $idx => &$item`?

**Review by aydun — COMMENTED** [review]
r-run: it works!

File: Civi/Api4/Generic/Traits/CustomValueActionTrait.php (hunk 1 above)
Comment (colemanw) [reply Civi/Api4/Generic/Traits/CustomValueActionTrait.php:56]: Actually no I think that would mess up the output values which uses the original `$items` not the rewritten keys.

Comment (aydun) [reply Civi/Api4/Generic/Traits/CustomValueActionTrait.php:56]: Ok

**Comment by mattwire** [comment]
@colemanw it looks like this bug could have been affecting other entities too such as address? Would it only affect them when called via api4 or more genericly?

**Comment by colemanw** [comment]
It would only affect APIv4 calls, which are still few & far-between in core. I also don't think the bug affects other entities; reading the code I don't see others that do extra processing where the formatting matters.

**Comment by eileenmcnaughton** [comment]
This seems mergeable & tested to me - added merge-ready in case @mattwire has any last comments

**Comment by mattwire** [comment]
Merging based on @eileenmcnaughton review. I just wanted to get clarification from @colemanw which he has done :-)

**Pull request merged**